### PR TITLE
Expose 'Reimport' on right-click context menu in the FileSystem panel

### LINKED ISSUE #6320: Expose 'Reimport' on right-click context menu in the FileSystem panel
### Describe the project you are working on

A 2D adventure game which uses importer plugins to generate Godot scenes and resources

### Describe the problem or limitation you are having in your project

When I'm working with import plugins that generate Godot scenes and resources (like vnen's Tiled importer among other custom plugins), sometimes I want to force a reimport to regenerate them. This could be because I'm iterating on the import code or because the generated files got into a bad state. Often I have the generated scene open in the editor and I want to reimport and reload the scene. So selecting the resource under Filesystem, then going to the import tab, then clicking reimport feels like an avoidable bit of navigation that could be sped up with a context menu item. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Right clicking the resource and clicking reimport cuts out navigation to different tabs. I'd be able to reimport my resources in two clicks in quick succession

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When the user right clicks an importable resource,  "Reimport" would appear in the right click menu, which the user could click to trigger a reimport without changing the current "Import As" setting.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I tried searching but I didn't find a way for a plugin to add entries to Filesystem's right-click menu

### Is there a reason why this should be core and not an add-on in the asset library?

This is to speed up a workflow and I'm not sure how to implement it as an add-on.

### PR DESCRIPTION
**Pull request by nongvantinh**
Implement proposal: [#6320](https://github.com/godotengine/godot-proposals/issues/6320)

*Bugsquad edit:*
- Closes https://github.com/godotengine/godot-proposals/issues/6320

### CODE REVIEW

File: editor/import_dock.cpp [hunk 1]
```diff
@@ -316,6 +316,22 @@ void ImportDock::set_edit_multiple_paths(const Vector<String> &p_paths) {
 	}
 }
 
+void ImportDock::reimport_resources(const Vector<String> &p_paths) {
+	switch (p_paths.size()) {
+		case 0:
+			ERR_FAIL_MSG("You need to select files to reimport them.");
+			break;
```
Comment (KoBeWi) [inline editor/import_dock.cpp:323]: ERR_FAIL already returns, so this break is not necessary (and unreachable).

**Comment by KoBeWi** [comment]
`_reimport_attempt()` for multiple files will reimport all files with the same type as the first selected file. You can e.g. select texture and audio file and clicking Reimport will trigger the warning about changing import type. If you accept, it will corrupt the imported files.

Either the function needs to be changed or `_reimport()` shouldn't appear if there is more than 1 type of resources selected.

**Comment by nongvantinh** [comment]
I understand. I tested the import tab in the current version and found that when I try to import two different file types, the editor doesn't let me reimport them. To maintain consistency, I'll try to implement the same functionality as the editor.

**Comment by KoBeWi** [comment]
The implementation can be much simpler:
```C++
bool resource_valid = true;
String main_extension;

for (int i = 0; i != p_paths.size(); ++i) {
	String extension = p_paths[i].get_extension();
	if (extension_list.has(extension)) {
		if (main_extension.is_empty()) {
			main_extension = extension
		} else if (extension != main_extension) {
			resource_valid = false;
			break;
		}
	} else {
		resource_valid = false;
		break;
	}
}
```

**Review by KoBeWi — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]
Seems useful.

**Comment by dogboydog** [comment]
Hey, thanks for implementing this. just tried this in the 4.2 dev3 snapshot. It seems like this doesn't work for all types?  For instance, when using the https://github.com/Kiamo2/YATI plugin, .tmj files  don't show Reimport 
image

This was actually the use case that I created the proposal for, to reimport my tilemaps easily. Another plugin I'm using, that I wrote myself, does have the Reimport option
image

Any idea why it wouldn't show up for some resources?

**Comment by KoBeWi** [comment]
Maybe there is something wrong with the YATI implementation, like missing function or something. Doesn't it print any errors when used? Can you show the code of your plugin?

**Comment by dogboydog** [comment]
I don't think I have errors from YATI but I can check later. This is my importer code which does show Reimport: 

https://github.com/dogboydog/YarnDonut/blob/godot4/addons/YarnDonut/Editor/YarnImporter.cs

**Comment by KoBeWi** [comment]
YATI returns "Node2D" in `GetResourceType()`, while the method expects Resource. I think it should be PackedScene in this case.

**Comment by dogboydog** [comment]
Thanks, I'll let the developer know

**Pull request merged**